src/string.cr

```diff
@@ -3015,6 +3112,20 @@ class String
     {bytes, bytesize}
   end
 
+  private def range_to_index_and_size(range)
+    from = range.begin
+    from += size if from < 0
+    raise IndexError.new if from < 0
```

Comment:
Perhaps it's not super useful but I think it's useful enough for a lot of cases, I often start to debug them by first printing the input values and boundaries, that step wouldn't be necessary then. But in general it also makes the language appear more friendly, at least it's *trying* to give as much information as possible, instead of throwing a blunt "wrong" into your face.

I started making all `IndexError`s verbose but got stuck on #2819, you can see the message I used here: https://github.com/jhass/crystal/blob/better_index_error/src/exception.cr#L189-L198